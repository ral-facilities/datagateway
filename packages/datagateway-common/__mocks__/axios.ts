import axios from 'axios';

const requests = {
  get: vi.fn(() => Promise.resolve({ data: {} })),
  post: vi.fn(() => Promise.resolve({ data: {} })),
  delete: vi.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  AxiosError: axios.AxiosError,
  isAxiosError: axios.isAxiosError,
};

export default requests;
