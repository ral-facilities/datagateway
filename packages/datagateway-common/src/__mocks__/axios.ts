// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

import axios from 'axios';

const requests = {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  AxiosError: axios.AxiosError,
};

export default requests;
