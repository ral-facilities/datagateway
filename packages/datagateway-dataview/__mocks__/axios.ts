import axios from 'axios';

const requests = {
  get: vi.fn((path) => {
    if (path === '/datagateway-dataview-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'Generic',
          idsUrl: 'ids',
          apiUrl: 'api',
          downloadApiUrl: 'download',
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: vi.fn(() => Promise.resolve({ data: {} })),
  delete: vi.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  isAxiosError: axios.isAxiosError,
};

export default requests;
