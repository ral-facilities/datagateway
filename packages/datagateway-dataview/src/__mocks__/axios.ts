// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

import axios from 'axios';

const requests = {
  get: jest.fn((path) => {
    if (path === '/datagateway-dataview-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'Generic',
          idsUrl: 'ids',
          apiUrl: 'api',
          downloadApiUrl: 'download',
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
  isAxiosError: axios.isAxiosError,
};

export default requests;
