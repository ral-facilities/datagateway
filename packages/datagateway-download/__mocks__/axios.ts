import axios from 'axios';

const requests = {
  get: vi.fn((path) => {
    if (path.includes('/topcat/user/cart/')) {
      return Promise.resolve({ data: { cartItems: [] } });
    } else if (path === '/datagateway-download-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'LILS',
          apiUrl: 'https://example.com/api',
          downloadApiUrl: 'https://example.com/downloadApi',
          idsUrl: 'https://example.com/ids',
          fileCountMax: 5000,
          totalSizeMax: 1000000000000,
          accessMethods: {
            https: {
              idsUrl: 'https://example.com/ids',
              displayName: 'HTTPS',
              description: 'Example description for HTTPS access method.',
            },
            globus: {
              idsUrl: 'https://example.com/ids',
              displayName: 'Globus',
              description: 'Example description for Globus access method.',
            },
          },
          routes: [
            {
              section: 'Test',
              link: '/download',
              displayName: 'Download',
              order: 0,
            },
          ],
        },
      });
    } else if (path.includes('/topcat/user/downloads')) {
      return Promise.resolve({ data: [] });
    } else {
      return Promise.resolve({ data: {} });
    }
  }),
  post: vi.fn(() => Promise.resolve({ data: {} })),
  put: vi.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};

export default requests;
