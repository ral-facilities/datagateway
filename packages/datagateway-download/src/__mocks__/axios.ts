// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

import axios from 'axios';

const requests = {
  get: jest.fn((path) => {
    if (path.includes('/topcat/user/cart/')) {
      return Promise.resolve({ data: { cartItems: [] } });
    } else if (path === '/datagateway-download-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'LILS',
          apiUrl: 'https://example.com/api',
          downloadApiUrl: 'https://example.com/downloadApi',
          idsUrl: 'https://example.com/ids',
          fileCountMax: 5000,
          totalSizeMax: 1000000000000,
          accessMethods: {
            https: {
              idsUrl: 'https://example.com/ids',
              displayName: 'HTTPS',
              description: 'Example description for HTTPS access method.',
            },
            globus: {
              idsUrl: 'https://example.com/ids',
              displayName: 'Globus',
              description: 'Example description for Globus access method.',
            },
          },
          routes: [
            {
              section: 'Test',
              link: '/download',
              displayName: 'Download',
              order: 0,
            },
          ],
        },
      });
    } else if (path.includes('/topcat/user/downloads')) {
      return Promise.resolve({ data: [] });
    } else {
      return Promise.resolve({ data: {} });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  put: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};

export default requests;
