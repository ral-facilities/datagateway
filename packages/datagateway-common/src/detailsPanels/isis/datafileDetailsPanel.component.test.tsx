import type { RenderResult } from '@testing-library/react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { UserEvent } from '@testing-library/user-event/dist/types/setup';
import axios from 'axios';
import * as React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { combineReducers, createStore } from 'redux';
import type { StateType } from '../../../lib';

import type { Datafile } from '../../app.types';
import dGCommonReducer from '../../state/reducers/dgcommon.reducer';
import DatafileDetailsPanel from './datafileDetailsPanel.component';

function renderComponent({
  rowData,
  detailsPanelResize,
}: {
  rowData: Datafile;
  detailsPanelResize?: () => void;
}): RenderResult {
  return render(
    <Provider
      store={createStore(
        combineReducers<Partial<StateType>>({ dgcommon: dGCommonReducer })
      )}
    >
      <QueryClientProvider client={new QueryClient()}>
        <DatafileDetailsPanel
          rowData={rowData}
          detailsPanelResize={detailsPanelResize}
        />
      </QueryClientProvider>
    </Provider>
  );
}

describe('Datafile details panel component', () => {
  let rowData: Datafile;
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
    rowData = {
      id: 1,
      name: 'Test 1',
      location: '/test/location',
      modTime: '2019-06-10',
      createTime: '2019-06-11',
      description: 'Test description',
    };
    axios.get = jest.fn().mockResolvedValue({
      data: [rowData],
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = renderComponent({ rowData });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show default tab on first render', async () => {
    renderComponent({
      rowData,
    });

    expect(
      await screen.findByRole('tab', { name: 'datafiles.details.label' })
    ).toHaveAttribute('aria-selected', 'true');
  });

  it('should let user switch between tabs', async () => {
    rowData.parameters = [
      {
        id: 2,
        stringValue: 'String test',
        type: {
          id: 3,
          name: 'String parameter',
          units: 'foo/s',
          valueType: 'STRING',
        },
      },
    ];

    renderComponent({ rowData });

    expect(
      await screen.findByRole('tabpanel', { name: 'datafiles.details.label' })
    ).toBeVisible();

    // switch tab
    await user.click(
      await screen.findByRole('tab', {
        name: 'datafiles.details.parameters.label',
      })
    );

    expect(
      await screen.findByRole('tabpanel', {
        name: 'datafiles.details.parameters.label',
      })
    ).toBeVisible();
  });

  it('should render parameters tab when present in the data', () => {
    rowData.parameters = [
      {
        id: 2,
        stringValue: 'String test',
        type: {
          id: 3,
          name: 'String parameter',
          units: 'foo/s',
          valueType: 'STRING',
        },
      },
      {
        id: 4,
        numericValue: 1337,
        type: {
          id: 5,
          name: 'Numeric parameter',
          units: 'bar/s',
          valueType: 'NUMERIC',
        },
      },
      {
        id: 6,
        dateTimeValue: '2019-09-10 11:48:00',
        type: {
          id: 7,
          name: 'Datetime parameter',
          units: 'baz/s',
          valueType: 'DATE_AND_TIME',
        },
      },
      {
        id: 8,
        dateTimeValue: '2019-09-10 11:48:00',
        type: {
          id: 9,
          name: 'Invalid parameter',
          units: 'n/a',
          valueType: '',
        },
      },
      {
        id: 10,
        stringValue: 'Missing PARAMETERTYPE',
      },
    ];

    const { asFragment } = renderComponent({
      rowData,
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call detailsPanelResize on load and when tabs are switched between', async () => {
    rowData.parameters = [
      {
        id: 2,
        stringValue: 'String test',
        type: {
          id: 3,
          name: 'String parameter',
          units: 'foo/s',
          valueType: 'STRING',
        },
      },
    ];

    const mockDetailsPanelResize = jest.fn();
    renderComponent({
      rowData,
      detailsPanelResize: mockDetailsPanelResize,
    });

    await waitFor(() => {
      expect(mockDetailsPanelResize).toHaveBeenCalledTimes(1);
    });

    await user.click(
      await screen.findByRole('tab', {
        name: 'datafiles.details.parameters.label',
      })
    );

    await waitFor(() => {
      expect(mockDetailsPanelResize).toHaveBeenCalledTimes(2);
    });
  });

  it('should not call detailsPanelResize if not provided', async () => {
    rowData.parameters = [
      {
        id: 2,
        stringValue: 'String test',
        type: {
          id: 3,
          name: 'String parameter',
          units: 'foo/s',
          valueType: 'STRING',
        },
      },
    ];
    const mockDetailsPanelResize = jest.fn();

    renderComponent({ rowData });

    await user.click(
      await screen.findByRole('tab', {
        name: 'datafiles.details.parameters.label',
      })
    );

    expect(mockDetailsPanelResize).not.toHaveBeenCalled();
  });

  it('should show "No <field> provided" incase of a null field', async () => {
    rowData = {
      id: 1,
      name: 'Test 1',
      modTime: '2019-06-10',
      createTime: '2019-06-11',
    };

    axios.get = jest.fn().mockResolvedValue({
      data: [rowData],
    });

    renderComponent({ rowData });

    expect(
      await screen.findByText('datafiles.details.description not provided')
    ).toBeInTheDocument();
  });

  it('should render datafile parameters tab and text "No parameters" when no data is present', async () => {
    rowData.parameters = [];
    axios.get = jest.fn().mockResolvedValue({
      data: [rowData],
    });

    renderComponent({ rowData });

    expect(
      await screen.findByText('datafiles.details.parameters.no_parameters')
    ).toBeInTheDocument();
  });
});
