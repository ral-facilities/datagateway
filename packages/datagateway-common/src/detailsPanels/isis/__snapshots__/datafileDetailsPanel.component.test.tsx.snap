// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datafile details panel component should render correctly 1`] = `
<DocumentFragment>
  <div
    id="details-panel"
    style="min-width: 0;"
  >
    <div
      class="MuiTabs-root css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-agdkjh-MuiTabs-scrollbarSize"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
        style="margin-bottom: 0px;"
      >
        <div
          aria-label="datafiles.details.tabs_label"
          class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-controls="datafile-details-panel"
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
            id="datafile-details-tab"
            role="tab"
            tabindex="0"
            type="button"
          >
            datafiles.details.label
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-11yukd5-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      aria-labelledby="datafile-details-tab"
      hidden=""
      id="datafile-details-panel"
      role="tabpanel"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-jddpo1-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            <b>
              Test 1
            </b>
          </h6>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-k7lt0k-MuiDivider-root"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            datafiles.details.description
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              Test description
            </b>
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            datafiles.details.location
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              /test/location
            </b>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Datafile details panel component should render parameters tab when present in the data 1`] = `
<DocumentFragment>
  <div
    id="details-panel"
    style="min-width: 0;"
  >
    <div
      class="MuiTabs-root css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-agdkjh-MuiTabs-scrollbarSize"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
        style="margin-bottom: 0px;"
      >
        <div
          aria-label="datafiles.details.tabs_label"
          class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-controls="datafile-details-panel"
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
            id="datafile-details-tab"
            role="tab"
            tabindex="0"
            type="button"
          >
            datafiles.details.label
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-controls="datafile-parameters-panel"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
            id="datafile-parameters-tab"
            role="tab"
            tabindex="-1"
            type="button"
          >
            datafiles.details.parameters.label
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-11yukd5-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      aria-labelledby="datafile-details-tab"
      hidden=""
      id="datafile-details-panel"
      role="tabpanel"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-jddpo1-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            <b>
              Test 1
            </b>
          </h6>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-k7lt0k-MuiDivider-root"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            datafiles.details.description
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              Test description
            </b>
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            datafiles.details.location
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              /test/location
            </b>
          </p>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="datafile-parameters-tab"
      hidden=""
      id="datafile-parameters-panel"
      role="tabpanel"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-jddpo1-MuiGrid-root"
        id="parameter-grid"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            String parameter
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              String test
            </b>
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            Numeric parameter
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              1337
            </b>
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
          >
            Datetime parameter
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            <b>
              2019-09-10
            </b>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
